firmware/lib/Gantry/Axis.h

```diff
@@ -1,69 +1,108 @@
+} AxisMotion;
+
+/**
+ * @enum VelSeg
+ * 
+ * @brief Enumeration of the segments of a trapezoidal velocity profile
+ */
 typedef enum {
-    VEL_SEG_ACCELERATE,
-    VEL_SEG_HOLD,
-    VEL_SEG_DECELERATE
-} VelocitySegment;
+    VEL_SEG_ACCELERATE,                //!< Accelerating up to the holding velocity
+    VEL_SEG_HOLD,                      //!< Staying constant at the holding velocity
+    VEL_SEG_DECELERATE                 //!< Decelerating down to a minimum velocity
+} VelSeg;
 
+/**
+ * @struct AxisState
+ * 
+ * @brief Specifies the current state of an axis
+ * 
+ * All fields are declared volatile since they may be modified from ISRs
+ */
 typedef struct {
-    uint32_t dir_pin, step_pin, encoder_pin_a, encoder_pin_b, ls_home, ls_far;
-    uint32_t acceleration, vel_max, vel_min;
-    Tc *timer;
-    uint32_t channel_velocity, channel_accel;
-    IRQn_Type irq_velocity, irq_accel;
-    void (*isr_encoder)(void);
-    void (*isr_limit_switch)(void);
-} AxisConfig;
+    volatile bool moving;              //!< true if the axis is currently moving
 
-typedef struct Axis
-{
-    uint32_t dir_pin;
-    uint32_t step_pin;
-    uint32_t accel;
-    uint32_t vel_min;
-    uint32_t vel_max;
-    uint32_t vel;
-    int32_t vel_profile_cur_trap[3] = {0, 0, 0}; // defined in counts accelerate, hold, decelerate
-    VelocitySegment tragectory_segment;
-    Encoder encoder;
-    LimitSwitch ls_home;
-    LimitSwitch ls_far_from_home;
-    // necessary timer info
-    Tc *timer;
-    uint32_t channel_velocity;
-    uint32_t channel_accel;
-    IRQn_Type irq_velocity;
-    IRQn_Type irq_accel;
-    Direction dir;
-    bool moving;
-    bool homing;
-    bool fault;
-} Axis;
+    volatile bool ls_home_pressed;     //!< true if the home limit switch is currently pressed
+    volatile bool ls_far_pressed;      //!< true if the far limit switch is currently pressed
 
-// setup functions
-void setup_axis(AxisConfig *axis_config, Axis *axis);
-void reset_axis(AxisConfig *axis_config, Axis *axis);
+    volatile uint32_t velocity;        //!< Current velocity of the axis
+    volatile uint32_t next_velocity;   //!< Pending velocity to set for the axis
+    volatile VelSeg velocity_segment;  //!< Current velcoity segment of the axis
+    volatile int32_t encoder_current;  //!< Current position of the axis in encoder counts
+    volatile int32_t encoder_target;   //!< Position at which the next segment transition will occur
+    volatile Direction dir;            //!< Current direction of motion of the axis
+} AxisState;
 
- // start/stop functions
-void start_axis(Axis *axis);
-void stop_axis(Axis *axis);
+/**
+ * @enum AxisResult
+ * 
+ * @brief Possible return values for a call to axis_start
+ */
+typedef enum {
+    AXIS_OK,                           //!< Axis movement started OK
+    AXIS_ERR_DIR_MISMATCH,             //!< Direction did not match count value signs
+    AXIS_ERR_ZERO_DIST,                //!< Total specified distance was zero
```

Comment:
Does specifying zero distance qualify as an error? Or should we just not move that axis.

Reply:
I guess there's nothing wrong with accepting a zero distance movement. I can switch it to not be an error